A43-sdk-authorization.md

```diff
@@ -0,0 +1,364 @@
+	  "scoped by services.",
+	"type": "object",
+	"properties": {
+		"name": {
+			"description": "The name of an authorization policy. It is mainly for monitoring and error message generation.",
+			"type": "string",
+		},
+		"deny_rules": {
+			"description": "List of deny rules to match. If a request matches any of the deny rules, then it will be denied. If none of"
+			  "the deny rules matches or there are no deny rules, the allow rules will be evaluated.",
+			"type": "array",
+			"items": {
+				"$ref": "#/definitions/rule",
+			},
+		},
+		"allow_rules": {
+			"description": "List of allow rules to match. The allow rules will only be evaluated after the deny rules. If a request"
+			  "matches any of the allow rules, then it will allowed. If none of the allow rules matches, it will be denied.",
+			"type": "array",
+			"items": {
+				"$ref": "#/definitions/rule",
+			},
+		}
+	},
+	"required": ["name", "allow_rules"]
+}
+```
+
+gRPC SDK authorization policy has a list of allow rules and a list of deny rules. 
+The following sequence is followed to make an authorization decision -
+1. Check for a match in deny rules, if a match is found, the request will be denied. 
+   If no match is found in deny rules, or there are no deny rules execute next step.
+2. Check for a match in allow rules, if a match is found, the request will be allowed. 
+   Note that allow rules is a required field here, if not present, policy is invalid.
+3. If no match is found in allow rules, we deny the request.
+
+In the following policy example
+- Peer identity from ["spiffe://foo.com/sa/admin1", "spiffe://foo.com/sa/admin2"] is 
+  authorized to access any RPC methods in pkg.service
+- Any authenticated user is allowed to access "foo" and "bar" RPC methods.
+- Nobody can access "secret" RPC method, not even the admins as deny rules are evaluated 
+  first.
+
+```json
+{
+	"name": "example-policy",
+	"allow_rules": [
+		{
+			"name": "admin-access",
+			"source": {
+				"principals": [
+					"spiffe://foo.com/sa/admin1",
+					"spiffe://foo.com/sa/admin2"
+				]
+			},
+			"request": {
+				"paths": ["/pkg.service/*"]
+			}
+		},
+		{
+			"name": "dev-access",
+			"source": {
+				"principals": [
+					"*"
+				]
+			},
+			"request": {
+				"paths": [
+					"/pkg.service/foo",
+					"/pkg.service/bar"
+				]
+			}
+		}
+	],
+	"deny_rules": [
+		{
+			"name": "deny-access",
+			"request": {
+				"paths": [
+					"*/secret"
+				]
+			}
+		}
+	]
+}
+```
+
+As mentioned previously, internally gRPC Authorization engine consumes Envoy RBAC 
+proto. The user provided JSON SDK authorization policy will be translated to Envoy 
+RBAC proto and provided to Engines for evaluation. Note that SDK authorization policy 
+is a subset of Envoy RBAC, and it does not support all the fields that are present in 
+Envoy RBAC.
+
+### API
+
+gRPC will support both static initialization and dynamically reloading the policy
```

Comment:
mTLS is a pre-requisite for using authz because it only supports `principals` in the policy. It should be mentioned here and also checked in the implementation?

Reply:
So we do check for authenticated connection when we are matching the principal values
https://github.com/grpc/grpc/blob/b9a643a81736d9efce5b0b6fce1cf6b1f66e1320/src/core/lib/security/authorization/matchers.cc#L177
I am not sure about mTLS being a pre-requisite for authz. Say we have allow_all kind policy (source and request empty), then it wouldn't check for whether connection is authenticated. It is only done when we use principal field.

Reply:
> we have allow_all kind policy

Sure, I didn't think of that but what's the value of such a policy?

Reply:
We could have allow_rules as below
allow_rules : [{"name": "allow-all"}]
Since source and request are empty, this would translate to an ALLOW RBAC policy with ANY permission and ANY principal.
This means, the server would allow all incoming requests. Without checking for TLS within engine.

However, we do assert for grpc_auth_context in C-core filter for ex. Which would exist only if connection is authenticated.

Reply:
> However, we do assert for grpc_auth_context in C-core filter for ex. Which would exist only if connection is authenticated.

So the only use of allow_all policy is to make sure the connection is mTLS aka authenticated. Otherwise not setting a policy is like allow_all. Is that right?

Reply:
Yes, that's my understanding.
For Java and Go, it depends on how the server is configured.
I am not sure what would happen if connection is not mTLS and we use SDK authz. I don't think the implementation checks for it. Will have to look into it.

@markdroth @ejona86 Should we explicitly mention that mTLS is a pre-requisite for using SDK authorization, and ensure that all three stacks check for this in their implementation?

Reply:
I think we should have mTLS as a pre-requisite for authz. It does not make sense to run the authz check against unauthenticated principles. Going back to your example, if we want to allow everything, why bother configuring authz in the first place?

Reply:
Let me take back the above comment. There could be use cases where a service owner may want to allow some RPC methods for ALL principles while restricting others to only authenticated principles. In Envoy RBAC, it also supports both authenticated and unauthenticated principles: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#envoy-v3-api-msg-config-rbac-v3-principal-authenticated

Reply:
I don't think there's any need to restrict to mTLS. If there is a principal in the rule, then that requires mTLS to match, but the rest of things still work without mTLS. Since other sources could be added in the future, there's nothing preventing non-mTLS usages becoming more useful in the future. And I agree having a way to let a few methods through via plaintext while requiring TLS for the rest is useful.

Reply:
Sounds good. So we will check for mTLS only when using principals field.
In which case, there won't be any difference in having SDK filter initialized with allow_all, or having no SDK filter, right?

Reply:
Thank you for the discussion. I agree we should check for TLS only with principal fields. For the other fields, TLS will not be a pre-requisite.